### PR TITLE
[LW] Relax conditions on start transaction flow

### PR DESCRIPTION
**Pull request by Jolyon-S**
**Goals (and why)**:
For whatever reason, the check that every start timestamp has a corresponding snapshot available fails surprisingly frequently on the odd internal service (yet appears fine for most). However, this is not a mandatory check: if it fails, it just means that some transactions cannot cache. Thus, instead of restarting the batch, just remove the constraint so that some of the transactions (but absolutely not all) will use no-op caches.

**Implementation Description (bullets)**:
* Remove constraint
* Fix a bug when removing state (note that a finally block was saving us in the production codepath)

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added a test that threw prior to this change.

**Concerns (what feedback would you like?)**:
Should we be concerned that we are semi-frequently hitting this issue in production at all?

**Where should we start reviewing?**:
Diff

**Priority (whenever / two weeks / yesterday)**:
Today - this was causing fairly significant issues internally!

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueScopingCacheImpl.java [hunk 1]
```diff
@@ -119,32 +118,31 @@ public synchronized void onSuccessfulCommit(long startTimestamp) {
         cache.finalise();
 
         Map<CellReference, CacheValue> cachedValues = cache.getValueDigest().loadedValues();
-        if (cachedValues.isEmpty()) {
-            return;
+        if (!cachedValues.isEmpty()) {
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueScopingCacheImpl.java:121]: drive-by fix: before, if there are no cached values, this wouldn't remove the state at the end! However, we did that as an extra step in a finally block in production, so this wasn't leaking memory anywhere outside of tests.

**Review by jeremyk-91 — COMMENTED** [review]
Discussed offline: while this is indeed safe in terms of correctness, the latest problem with an afflicted stack may not actually relate to this code (and I'm not sure it is always the preferred option to return a no-op cache if we do get into this state).

**Review by jeremyk-91 — APPROVED** [review]
👍  thanks, agree this version is 100% safe.
